### PR TITLE
Track Expense flow

### PR DESCRIPTION
**Pull request by shubham1206agra**
### Fixed Issues

$ https://github.com/Expensify/App/issues/36362

#### Test 1

1. Go to Self DM report.
2. Click on + Button.
3. Select Track Expense.
4. Enter the amount (Test using different currencies too).
5. Click on Next (Confirm the amount shown in the confirmation page matches).
6. [Optional] Click on 3 dots > Add receipt to add a receipt.
7. Click on Track Expense at the confirmation step.

#### Test 2

1. Go to Self DM report.
2. Click on + Button.
3. Select Track Expense.
4. Go to Scan tab.
5. Upload a receipt.
6. Click on Track Expense at the confirmation step.

#### Test 3

1. Click on FAB.
2. Select Track Expense.
3. Enter the amount (Test using different currencies too).
4. Click on Next (Confirm the amount shown in the confirmation page matches).
5. [Optional] Click on 3 dots > Add receipt to add a receipt.
6. Click on Track Expense at the confirmation step.

#### Test 4

1. Click on FAB Button.
2. Select Track Expense.
3. Go to Scan tab.
4. Upload a receipt.
5. Click on Track Expense at the confirmation step.

#### Test 5 (Delete Track Expense)

1. Go to the Track Expense you created.
2. Press on 3 dot menu.
3. Click on Delete expense.
4. Confirm the delete.
5. Verify that the track expense is deleted and you are navigated to self DM.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests

### QA Steps

Same as Tests.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Screenshots for Design Reference</summary>

Screenshot 2024-03-20 at 6 23 15 AM
Screenshot 2024-03-20 at 6 23 28 AM
Screenshot 2024-03-20 at 6 23 44 AM
Screenshot 2024-03-20 at 6 23 53 AM
Screenshot 2024-03-20 at 6 25 16 AM
Screenshot 2024-03-20 at 6 25 21 AM
Screenshot 2024-03-20 at 6 25 33 AM
Screenshot 2024-03-20 at 6 25 59 AM
Screenshot 2024-03-20 at 6 26 50 AM
Screenshot 2024-03-20 at 6 27 07 AM

</details>

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/58412969/49b50de4-179f-4e48-81e7-64941cbf531f

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/58412969/61f7d585-1a7e-4550-a0cf-566413d70f2e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/58412969/d84f2d5b-4aa7-4b3e-a9ab-27ff7af36d4e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/58412969/183d496e-744f-41b6-8ec2-ef3ab276b3fc

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/58412969/e5a931ef-7d80-4e0a-9f0f-815105eff954

https://github.com/Expensify/App/assets/58412969/2e65832b-02a5-43c1-b895-bd3734848e7e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/58412969/52e6a17e-036b-4f1d-9981-bd7725b65e52

</details>

### CODE REVIEW

File: src/CONST.ts [hunk 1]
```diff
@@ -1263,6 +1265,7 @@ const CONST = {
             SEND: 'send',
             SPLIT: 'split',
             REQUEST: 'request',
+            TRACK_EXPENSE: 'track-expense',
```
Comment (thienlnam) [inline src/CONST.ts:1268]: Let's do underscore since it's more common `track_expense`

File: src/pages/iou/request/IOURequestStartPage.js [hunk 2]
```diff
@@ -103,7 +104,7 @@ function IOURequestStartPage({
 
     const isExpenseChat = ReportUtils.isPolicyExpenseChat(report);
     const isExpenseReport = ReportUtils.isExpenseReport(report);
-    const shouldDisplayDistanceRequest = isExpenseChat || isExpenseReport || isFromGlobalCreate;
+    const shouldDisplayDistanceRequest = isExpenseChat || isExpenseReport || isFromGlobalCreate || iouType === CONST.IOU.TYPE.TRACK_EXPENSE;
```
Comment (thienlnam) [inline src/pages/iou/request/IOURequestStartPage.js:107]: We will want to hide the distance tab in the track expense selector for now - for context it's in the process of being refactored to fit the IOU flow

**Review by thienlnam — COMMENTED** [review]
Thanks, just a couple notes for far - it looks the track flow doesn't yet auto-select the self-DM as destination?

File: src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js [hunk 3]
```diff
@@ -179,13 +188,28 @@ function FloatingActionButtonAndPopover(props) {
...
+                              {
+                                  icon: Expensicons.TrackExpense,
+                                  text: 'Track Expense',
+                                  onSelected: () =>
+                                      interceptAnonymousUser(() =>
+                                          IOU.startMoneyRequest_temporaryForRefactor(
+                                              CONST.IOU.TYPE.TRACK_EXPENSE,
+                                              // When starting to create a track expense from the global FAB, we need to retrieve selfDM reportID.
+                                              // If it doesn't exist, we generate a random optimistic reportID and use it for all of the routes in the creation flow.
+                                              props.account.selfDMReportID || ReportUtils.generateReportID(),
```
Comment (thienlnam) [inline src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js:202]: Actually, can we instead add a method in ReportUtils that will get the selfDM report? We can't add the onyx key until all users are given the self-DM, and even then if we can get this reportID on the front-end from the reports in onyx, that is preferred

**Review by thienlnam — CHANGES_REQUESTED** [review]
Couple comments so far, also looks like we are failing some ts typechecks

File: src/libs/ReportUtils.ts [hunk 4]
```diff
@@ -4349,6 +4349,10 @@ function getMoneyRequestOptions(report: OnyxEntry<Report>, policy: OnyxEntry<Pol
     const hasMultipleOtherParticipants = otherParticipants.length > 1;
     let options: Array<ValueOf<typeof CONST.IOU.TYPE>> = [];
 
+    if (isSelfDM(report)) {
+        options = [CONST.IOU.TYPE.TRACK_EXPENSE];
+    }
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/ReportUtils.ts:4380]: Ah no, from the FAB flow we always track directly to the self-DM

When we track this from the workspace chat, it automatically selects that participant

File: src/libs/ReportUtils.ts [hunk 5]
```diff
@@ -4482,6 +4498,10 @@ function getMoneyRequestOptions(report: OnyxEntry<Report>, policy: OnyxEntry<Pol
         options = [CONST.IOU.TYPE.SPLIT];
     }
 
+    if (canUseTrackExpense && isPolicyExpenseChat(report) && report?.isOwnPolicyExpenseChat) {
+        options = [...options, CONST.IOU.TYPE.TRACK_EXPENSE];
+    }
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/ReportUtils.ts:4503]: Sure that works, we can create another issue for this to get added into the workspace

**Comment by thienlnam** [comment]
Crashing trying to use Track with
```
startMoneyRequest_temporaryForRefactor is not a function

```
Screenshot 2024-03-18 at 2 43 57 PM

**Review by ishpaul777 — COMMENTED** [review]
Couple comments so far.. Reviewed 70 %

File: src/libs/actions/IOU.ts [hunk 6]
```diff
@@ -1002,6 +1188,137 @@ function getMoneyRequestInformation(
...
+    // STEP 1: Get existing chat report
+    let chatReport = !isEmptyObject(parentChatReport) && parentChatReport?.reportID ? parentChatReport : null;
+
+    // The chatReport always exist and we can get it from Onyx if chatReport is null.
+    if (!chatReport) {
+        chatReport = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${participant.reportID}`] ?? null;
+    }
+
+    // If we still don't have a report, it likely doens't exist and we will early return here as it should not happen
+    // Maybe later, we can build an optimistic selfDM chat.
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/actions/IOU.ts:1221]: It's not guaranteed - I would just add this comment
// Self-DMs should exist for most people as it is created on re-authentication or new account creation. However, in the scenario where an existing user authenticates and the migration job has not finished running yet, there will be a short amount of time until the self-DM does not exist

File: src/libs/actions/IOU.ts [hunk 7]
```diff
@@ -1002,6 +1188,137 @@ function getMoneyRequestInformation(
...
+    // STEP 3: Build optimistic receipt and transaction
+    const receiptObject: Receipt = {};
+    let filename;
+    if (receipt?.source) {
+        receiptObject.source = receipt.source;
+        receiptObject.state = receipt.state ?? CONST.IOU.RECEIPT_STATE.SCANREADY;
+        filename = receipt.name;
+    }
+    const existingTransaction = allTransactionDrafts[`${ONYXKEYS.COLLECTION.TRANSACTION_DRAFT}${CONST.IOU.OPTIMISTIC_TRANSACTION_ID}`];
+    const isDistanceRequest = existingTransaction && existingTransaction.iouRequestType === CONST.IOU.REQUEST_TYPE.DISTANCE;
```
[the remark this answers is not in the document]
Comment (thienlnam) [reply src/libs/actions/IOU.ts:1239]: Yes, we are working on some BE changes to allow distance requests here https://expensify.slack.com/archives/C066HJM2CAZ/p1710796683487119, we'll get back to this later

**Comment by thienlnam** [comment]
Seeing new crashes for
```
ReportUtils.ts:2147 Uncaught ReferenceError: reportAction is not defined
    at Module.getReportPreviewMessage (ReportUtils.ts:2147:21)
    at Module.getLastMessageTextForReport (OptionsListUtils.ts:443:46)
    at Object.getOptionData (SidebarUtils.ts:269:35)
    at eval (OptionRowLHNData.tsx:44:28)
    at mountMemo (react-dom.development.js:17091:19)
    at Object.useMemo (react-dom.development.js:17564:16)
    at useMemo (react.development.js:1651:21)
    at OptionRowLHNData (OptionRowLHNData.tsx:42:27)
    at renderWithHooks (react-dom.development.js:16175:18)
    at updateFunctionComponent (react-dom.development.js:20382:20)
```

**Comment by shawnborton** [comment]
Gonna spin up a test build so @Expensify/design can have a looksie.

**Comment by dubielzyk-expensify** [comment]
Do I need to do anything special? I don't see the Track option in the adhoc build

**Comment by shawnborton** [comment]
I think it was just `trackExpense`

**Review by ishpaul777 — APPROVED** [review]
LGTM!

whats left:

1. Delete expense not workin properly 
2. pusher update not working and infinite loading for the expense preview
3. not able to edit a tracked expense

**Review by thienlnam — APPROVED** [review]
Awesome, thanks - there's a few things that will need BE updates. I'm planning on making issues so we can track those but it doesn't need to block this PR.

**Comment by thienlnam** [comment]
@shawnborton / @dubielzyk-expensify The beta takes an hour to kick in, are you able to see the option now?

**Comment by shawnborton** [comment]
Still no. Are you able to check/add me to the beta in case I messed it up?

**Comment by thienlnam** [comment]
Ah yeah that is currently known 
https://github.com/Expensify/App/pull/37282#pullrequestreview-1949415677
Editing and deleting need some additional changes in the BE

**Comment by thienlnam** [comment]
Merging so we can get this project back on track for the may release - please let us know if there are other design items that look off. We're keeping the beta on until we get these BE issues fixed

**Comment by ikevin127** [comment]
Because `optimisticData` related to the `ONYXKEYS.NVP_QUICK_ACTION_GLOBAL_CREATE` onyx key was missing from the `buildOnyxDataForTrackExpense`, this caused the quick action to not update while offline (optimistically):
- https://github.com/Expensify/App/issues/40097

File: src/libs/ReportUtils.ts [hunk 8]
```diff
@@ -1227,6 +1236,18 @@ function isIOURequest(report: OnyxEntry<Report>): boolean {
...
 }
 
+/**
+ * A Track Expense Report is a thread where the parent the parentReportAction is a transaction, and
+ * parentReportAction has type of track.
+ */
+function isTrackExpenseReport(report: OnyxEntry<Report>): boolean {
+    if (isThread(report)) {
+        const parentReportAction = ReportActionsUtils.getParentReportAction(report);
+        return !isEmptyObject(parentReportAction) && ReportActionsUtils.isTrackExpenseAction(parentReportAction);
```
Comment (suneox) [inline src/libs/ReportUtils.ts:1246]: Coming from #52240 BZ checklist: 

After update track expense to transaction, the type of parent action still is `track` so we have to check parentReportID with selfDMReportID to detect it's a track expense

**Pull request merged**